### PR TITLE
lock now fixes

### LINKED ISSUE #305: Hitting "Lock Now" means the app is always locked
Prerequisites:

* Signed in on Lockbox
* Start in "Unlocked" state

Steps (all performed in less than `autlock` time):

1.  click "Lock Now" from drawer
2. click "Unlock" (follow unlock process)
3. hit `back` button
  * taken to previous activity (usually Home)
4. open Lockbox

Expected
* Lockbox is unlocked

Actual
* Lockbox is locked

Device:
* Pixel 3XL
* Android 9

### LINKED ISSUE #327: Stuck in a loop and unable to unlock
(This seems to only be an issue with Device Auth screen, Fingerprint dialog works on 1st attempt)

Steps to reproduce:
Select “Lock Now” from nav drawer
Tap “Unlock” button on Lock screen
Enter PIN
Upon completion, brings up PIN screen again
Sometimes repeats

Requires #260

### PR DESCRIPTION
**Pull request by sashei**
fixes #305 

## Testing Notes

I _believe_ this resolves ~~two~~ just one autolocking issue; ~~the first in which the user gets stuck in an unlocking loop and is unable to unlock the app at all, and the second~~ as detailed in #305. I would love to get some more eyes / hands to sanity check me though!

## To Do

- [x] add “WIP” to the PR title if pushing up but not complete nor ready for review
- [x] double check the original issue to confirm it is fully satisfied
- [x] add testing notes and screenshots in PR description to help guide reviewers
- [x] add unit tests
- [x] make sure CI builds are passing (e.g.: fix lint and other errors)

### CODE REVIEW

**Comment by sashei** [comment]
I thought I had concrete steps to reproduce #327 but I haven't been able to now >.< I _think_ this solves for that case but I would hesitate to close the issue without anything to test against.

File: app/src/main/java/mozilla/lockbox/store/AutoLockStore.kt [hunk 1]
```diff
@@ -84,7 +88,7 @@ class AutoLockStore(
     }
 
     private fun autoLockTimeElapsed(): Boolean {
-        val autoLockTimerDate = preferences.getLong(Constant.Key.autoLockTimerDate, -1)
+        val autoLockTimerDate = preferences.getLong(Constant.Key.autoLockTimerDate, Long.MAX_VALUE)
```
[the remark this answers is not in the document]
Comment (jhugman) [reply app/src/main/java/mozilla/lockbox/store/AutoLockStore.kt]: :+1:

File: app/src/main/java/mozilla/lockbox/store/AutoLockStore.kt [hunk 2]
```diff
@@ -49,14 +48,18 @@ class AutoLockStore(
             .map { lockCurrentlyRequired() }
             .subscribe(lockRequired as Subject)
 
-        Observables.combineLatest(lifecycleStore.lifecycleEvents, dataStore.state)
-            .filter { it.first == LifecycleAction.Background && it.second != DataStore.State.Locked }
+        lifecycleStore.lifecycleEvents
+            .filter { it == LifecycleAction.Background }
+            .switchMap { dataStore.state.take(1) }
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/store/AutoLockStore.kt]: I am concerned that this store is becoming too complicated for the thing it's doing. 

Conceptually, I feel it should be emitting events when the auto-lock timer is triggering, and the route presenter should be mapping them into lock actions.

By shoving all this logic here, we're spreading the auto-locking/locking/routing into lots of places which will make things harder to debug in the long run.

File: app/src/main/java/mozilla/lockbox/store/AutoLockStore.kt [hunk 3]
```diff
@@ -65,10 +68,11 @@ class AutoLockStore(
 
     override fun injectContext(context: Context) {
         preferences = PreferenceManager.getDefaultSharedPreferences(context)
+        (lockRequired as Subject).onNext(lockCurrentlyRequired())
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/store/AutoLockStore.kt]: `lockRequired` should perhaps be a `PublishSubject` called `autolockTriggered`. We discussed this in person, but I don't know who should do the work — here or another PR. :)

**Pull request merged**